Fix concurrent search and index delete

henningandersen opened the pull request:
Changed order of listener invocation so that we notify before
registering search context and notify after unregistering same.

This ensures that count up/down like what we do in ShardSearchStats
works. Otherwise, we risk notifying onFreeScrollContext before notifying
onNewScrollContext (same for onFreeContext/onNewContext, but we
currently have no assertions failing in those).

Closes #28053

henningandersen committed:
Fix concurrent search and index delete

Changed order of listener invocation so that we notify before
registering search context and notify after unregistering same.

This ensures that count up/down like what we do in ShardSearchStats
works. Otherwise, we risk notifying onFreeScrollContext before notifying
onNewScrollContext (same for onFreeContext/onNewContext, but we
currently have no assertions failing in those).

elasticmachine commented:
Pinging @elastic/es-search

elasticmachine commented:
Pinging @elastic/es-distributed

s1monw commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -550,12 +551,19 @@ final SearchContext createAndPutContext(ShardSearchRequest request) throws IOExc
                 context.indexShard().getSearchOperationListener().onNewScrollContext(context);
             }
             context.indexShard().getSearchOperationListener().onNewContext(context);
+            putContext(context);
```
I don't understand why you moved the `putContext(context)`

henningandersen replied:
By registering the search context in `activeContexts` after having invoked `onNewContext/onNewScrollContext`, we guarantee that for a specific `SearchContext`, the call to `onNewXXX` happens before the matching call the `onFreeXXX`.

s1monw replied:
I see I didn't think about `SearchService.afterIndexRemoved`.  my issue here is that we have to call  `onFreeXXX` but if one of the `onNewContext /onNewScrollContext` fails we don't register and fail? I think we need extra protection for this?

s1monw commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -550,12 +551,19 @@ final SearchContext createAndPutContext(ShardSearchRequest request) throws IOExc
+            // ensure that if index is deleted concurrently, we free the context immediately, either here or in afterIndexRemoved
+            try {
+                indicesService.indexServiceSafe(request.shardId().getIndex());
+            } catch (IndexNotFoundException e) {
```
I don't think you need the catch clause here - we free [below](https://github.com/elastic/elasticsearch/pull/42621/files#diff-a42a639093633662ea658bec29044a4cR571)

henningandersen replied:
How did I miss that, thanks! Fixed in 8edda7b36f7009f6d01d44fabe97e94153d7d06e

s1monw replied:
but is this necessary or can we just rearrange the registration?

s1monw reviewed (commented):
I am not following what the issue is here and why we don't clean it up correctly today? Can you show me what fails here in what situation. Sorry for being slow...

henningandersen commented:
Thanks for reviewing, @s1monw . The problem we were facing is that `onFreeScrollContext` could be called prior to `onNewScrollContext` for a specific `SearchContext`. This in turn makes `totalStats.scrollCurrent` in `ShardSearchStats` go negative, tripping the assertion in `ShardSearchStats.onFreeScrollContext`. Intuitively, it also seems unexpected for a `SearchOperationListener` to be exposed to this order.

The wrong order of invocation could occur in following scenario:
1. Someone does a search and a SearchContext is created and put into `SearchService.activeContexts` (but listeners not invoked yet).
2. The index is deleted and `SearchService.afterIndexRemoved` is called, which frees all `activeContexts`, thus invoking the listeners too.
3. Now the original thread resumes and invokes 'onNewContext/onNewScrollContext`.

henningandersen committed:
Removed duplicate catch block.

henningandersen committed:
Always call onFree plus test fix

s1monw commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -548,19 +548,35 @@ final SearchContext createAndPutContext(ShardSearchRequest request) throws IOExc
             putContext(context);
-            if (request.scroll() != null) {
+            // ensure that if index is deleted concurrently, we free the context immediately, either here or in freeAllContextForIndex
+            indicesService.indexServiceSafe(request.shardId().getIndex());
```
I am still not sure why we do this here. Why can't we just let the search go and fail whenever?

henningandersen replied:
It might be overkill... The way I understand the code, the primary purpose of `afterIndexRemoved` is to ensure that we free memory and file system resources as quickly as possible when an index is deleted/closed. Here we try to ensure the same happens also during a race condition. Without this, we theoretically risk holding on to memory/file system resources for a while after the index is deleted/closed. With this, we should free such resources as soon as the index is deleted (and any current search phases have run to completion).

I guess there is a trade-off in that doing this double validation has a small cost and the likelihood of seeing the race condition is very small. I can certainly remove this double validation if you think?

s1monw replied:
I think we should remove the double validation. But the fix you added in this PR is crucial!

henningandersen replied:
Thanks @s1monw , I have removed the double validation, please have another look at your convenience.

henningandersen committed:
Close search context if onNewXXX fails.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_concurrent_search_index_delete

henningandersen committed:
Removed double check for index delete.

Removed double check for index delete to avoid the performance overhead.
Removed validation of store ref count, since IndexService periodically
does a shard refresh, which disturbs this validation.

s1monw commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -648,18 +664,22 @@ private void freeAllContextForIndex(Index index) {
         }
     }
 
+    private void onFreeContext(SearchContext context) {
```
can you add an assertion here that ensures that we removed the context. I think it's trappy that we have an this method and we need to make sure that we never call if if we haven't removed the context. The special case is fina

s1monw reviewed (approved):
LGTM left one comment

henningandersen committed:
Assert that context is inactive when freeing.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_concurrent_search_index_delete

henningandersen commented:
Backported to 7.x, 7.2.
Backport to 6.8 pending.
